crypto/core_fetch.c

```diff
@@ -30,7 +30,31 @@ static void ossl_method_construct_this(OSSL_PROVIDER *provider,
 {
     struct construct_data_st *data = cbdata;
     void *method = NULL;
+    int global_store = data->force_store || !no_store;
 
+    /*
+     * Check to see if a method has already been constructed from this
+     * implementation.  It's pointless to do this check if we're not
+     * supposed to store the method in the global store.
+     *
+     * Note that while data->mcm->put() takes a property *definition*,
+     * data->mcm->get() takes a property *query*, but we're using the
+     * property *definition* string as a property query here.
+     * This is possible because property definitions have a syntax that
+     * can also be used for a property query with equality tests.  We do
+     * this to look up already existing methods matching the implementation
+     * that we just got from the provider (via |algo|).
+     */
+    if (global_store
+        && (method = data->mcm->get(data->libctx, NULL, data->operation_id,
+                                    algo->algorithm_name,
+                                    algo->property_definition,
+                                    data->mcm_data)) != NULL)
+        goto end;
```

Comment:
So if the method exists in the global store, we are not going to add it to the local store, right?
But why are we adding it to the local store at all?

Reply:
We're adding it to a local store so `ossl_method_construct()` can find it after we've added all methods that the providers had to offer (see (new) line 117).  This fullfils two objectives: 1) only do a second lookup on stuff that is newly added, 2) do a second lookup on methods that aren't supposed to remain in any store.

Reply:
but when the method happens to exist in the global store it will not be added
to the local store which mitght be a different one ?

Reply:
If it does, it was already picked up by `ossl_method_construct`.

Reply:
So why don't you add the method from the global store to the local store, so ossl_method_construct can
pick it up?

Reply:
or does this execute under some kind of global mutex?

Reply:
> So why don't you add the method from the global store to the local store, so `ossl_method_construct` can pick it up?

Like I said, `ossl_method_construct` has already picked it up.  It's the first thing it tries to do, and it only calls `ossl_method_construct_this` via `ossl_algorithm_do_all` when the implementation isn't in the global store at that time.  It's pointless to copy methods from the global store to the temporary store after that.

However, it's true that this isn't exactly thread safe.  In other words, we currently expact `FOO_fetch()` to be called in a single thread...  I'm unsure if we're explicit about that, though...  I think that's worth a separate issue if it turns out that we should be more protective.